**Theme:**

*“Most deep-learning failures are just bad learning rates.”*

# Imports

In [1]:
import torch 
import torch.nn as nn

torch.__version__

'2.8.0+cu129'

In [2]:
from torchvision import datasets, transforms 
from torch.utils.data import DataLoader, random_split   

# Optimizer, Learning rates and Why training fails

## Try with a high learning rate (Break the model)

In [3]:
transform = transforms.ToTensor() 

train_data_full = datasets.MNIST(root='data', train=True, download=True, transform=transform)  

train_size = int(0.8 * len(train_data_full))   
val_size = len(train_data_full) - train_size 

train_data, val_data = random_split(train_data_full, [train_size, val_size])  

train_loader = DataLoader(train_data, batch_size=64, shuffle=True) 
val_loader = DataLoader(val_data, batch_size=64, shuffle=False) 

In [4]:
# Neural Network
class MLP(nn.Module):
    def __init__(self):
        super().__init__()
        self.net = nn.Sequential(
            nn.Flatten(),
            nn.Linear(28*28, 128),
            nn.ReLU(),
            nn.Linear(128,10),
        )
    
    def forward(self, x):
        return self.net(x) 

model = MLP()
loss_fn = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=1.0)  

📌 Observe:

- Train loss

- Validation loss

Write what happens.

In [6]:
for epoch in range(10):
    # TRAIN 
    model.train()
    train_loss = 0.0 
    for images, labels in train_loader:
        # forward pass 
        outputs = model(images) 
        # compute loss 
        loss = loss_fn(outputs, labels) 

        # zero gradient 
        optimizer.zero_grad() 
        # backward pass
        loss.backward() 
        # update parameters 
        optimizer.step()

        train_loss += loss.item() 

    # VALIDATION 
    model.eval() 
    val_loss = 0.0 
    with torch.no_grad():
        for images, labels in val_loader:
            outputs = model(images) 
            loss = loss_fn(outputs, labels) 

            val_loss += loss.item() 

    print(f"Epoch: {epoch+1} | Train Loss: {train_loss/len(train_loader):.4f} | Validation Loss: {val_loss/len(val_loader):.4f}")

Epoch: 1 | Train Loss: 2.3297 | Validation Loss: 2.3782
Epoch: 2 | Train Loss: 2.3694 | Validation Loss: 2.3606
Epoch: 3 | Train Loss: 2.3664 | Validation Loss: 2.3970
Epoch: 4 | Train Loss: 2.3747 | Validation Loss: 2.3571
Epoch: 5 | Train Loss: 2.3703 | Validation Loss: 2.3924
Epoch: 6 | Train Loss: 2.3747 | Validation Loss: 2.3967
Epoch: 7 | Train Loss: 2.3647 | Validation Loss: 2.3909
Epoch: 8 | Train Loss: 2.3702 | Validation Loss: 2.4049
Epoch: 9 | Train Loss: 2.3760 | Validation Loss: 2.3614
Epoch: 10 | Train Loss: 2.3710 | Validation Loss: 2.3594


## Try with a very low learning rate

In [17]:
# Model, loss function, optimizer
model2 = MLP()
loss_fn2 = nn.CrossEntropyLoss()
optimizer2 = torch.optim.Adam(model2.parameters(), lr=0.00001)  

📌 Observe:

- How fast does loss decrease?

- Compare with lr = 0.001

In [18]:
for epoch in range(10):
    # TRAIN 
    model2.train()
    train_loss = 0.0 
    for images, labels in train_loader:
        # forward pass 
        outputs = model2(images) 
        # compute loss 
        loss = loss_fn2(outputs, labels) 

        # zero gradient 
        optimizer2.zero_grad() 
        # backward pass
        loss.backward() 
        # update parameters 
        optimizer2.step()

        train_loss += loss.item() 

    # VALIDATION 
    model2.eval() 
    val_loss = 0.0 
    with torch.no_grad():
        for images, labels in val_loader:
            outputs = model2(images) 
            loss = loss_fn2(outputs, labels) 

            val_loss += loss.item() 

    print(f"Epoch: {epoch+1} | Train Loss: {train_loss/len(train_loader):.4f} | Validation Loss: {val_loss/len(val_loader):.4f}")

Epoch: 1 | Train Loss: 2.1156 | Validation Loss: 1.8907
Epoch: 2 | Train Loss: 1.6696 | Validation Loss: 1.4601
Epoch: 3 | Train Loss: 1.2920 | Validation Loss: 1.1420
Epoch: 4 | Train Loss: 1.0288 | Validation Loss: 0.9276
Epoch: 5 | Train Loss: 0.8517 | Validation Loss: 0.7818
Epoch: 6 | Train Loss: 0.7296 | Validation Loss: 0.6795
Epoch: 7 | Train Loss: 0.6433 | Validation Loss: 0.6062
Epoch: 8 | Train Loss: 0.5804 | Validation Loss: 0.5515
Epoch: 9 | Train Loss: 0.5331 | Validation Loss: 0.5097
Epoch: 10 | Train Loss: 0.4965 | Validation Loss: 0.4770


## Compare Optimizers

In [21]:
model3 = MLP()

optimizer3 = torch.optim.SGD(model3.parameters(), lr=0.1) # now with SGD instead of Adam

In [22]:
for epoch in range(10):
    # TRAIN 
    model3.train()
    train_loss = 0.0 
    for images, labels in train_loader:
        # forward pass 
        outputs = model3(images) 
        # compute loss 
        loss = loss_fn2(outputs, labels) 

        # zero gradient 
        optimizer3.zero_grad() 
        # backward pass
        loss.backward() 
        # update parameters 
        optimizer3.step()

        train_loss += loss.item() 

    # VALIDATION 
    model3.eval() 
    val_loss = 0.0 
    with torch.no_grad():
        for images, labels in val_loader:
            outputs = model3(images) 
            loss = loss_fn2(outputs, labels) 

            val_loss += loss.item() 

    print(f"Epoch: {epoch+1} | Train Loss: {train_loss/len(train_loader):.4f} | Validation Loss: {val_loss/len(val_loader):.4f}")

Epoch: 1 | Train Loss: 0.4768 | Validation Loss: 0.2785
Epoch: 2 | Train Loss: 0.2433 | Validation Loss: 0.2154
Epoch: 3 | Train Loss: 0.1882 | Validation Loss: 0.1763
Epoch: 4 | Train Loss: 0.1527 | Validation Loss: 0.1513
Epoch: 5 | Train Loss: 0.1274 | Validation Loss: 0.1377
Epoch: 6 | Train Loss: 0.1094 | Validation Loss: 0.1200
Epoch: 7 | Train Loss: 0.0950 | Validation Loss: 0.1135
Epoch: 8 | Train Loss: 0.0842 | Validation Loss: 0.1106
Epoch: 9 | Train Loss: 0.0755 | Validation Loss: 0.1024
Epoch: 10 | Train Loss: 0.0676 | Validation Loss: 0.1003


- Which converges faster?

- Which is more stable?

---

Explain:

- What happens if learning rate is too high?

- What happens if it is too low?

---

Many people blame:

- Model

- Data

- rchitecture

**But 80% of training failures are just wrong learning rates.**

# Interaction

### our results are spot-on ✅


`lr = 1.0`

- Loss stuck around `~2.3`

- ➡ That’s ≈ log(10) → random guessing

- Meaning: optimizer is jumping around, not learning.

`lr = 0.00001`

- Loss decreases, but painfully slow
- ➡ Model can learn, but training is inefficient.

Adam vs SGD

- Despite higher LR, Adam worked decently
- ➡ Because Adam internally adapts the learning rate per parameter, while SGD relies purely on our LR choice.

---

### The professional default (industry rule)



For most problems, start with:
```python
optimizer = Adam(model.parameters(), lr=1e-3)
```
Why?

- Works on most datasets

- Robust to bad initialization

- Requires minimal tuning

This is why Adam is used in 90% of research code.

---

### Learning rate finding (practical method) 

#### Method 1 — Try 3 values (most common)

Always try:

| Learning rate | Meaning       |
| ------------- | ------------- |
| `1e-1`        | Too high test |
| `1e-3`        | Safe default  |
| `1e-5`        | Too low test  |


Pick the one that:

- Decreases loss fastest

- Is stable

We already did this instinctively. That’s good.

---

#### Method 2 — LR Range Test (advanced, optional) 

Gradually increase LR during training:

- Track loss

- Choose LR just before loss explodes

Used in:

- fast.ai

- Large-scale training

We’ll learn this later.

### Choosing the optimizer (clear rules)


Use Adam when:

- Dataset is small/medium

- Features are noisy

- We want fast convergence

- We are prototyping

Use SGD (+ momentum) when:

- Dataset is very large

- We care about final generalization

- We are training CNNs from scratch

> Interview answer:

Adam converges faster, SGD often generalizes better.

---

### Why SGD worked with lr = 0.1

SGD update is simple:

$𝑤 = 𝑤 − \eta\cdot∇𝑤$

MNIST is:

- Simple

- Well-conditioned

- Clean

So SGD behaves well.

**On messy data, Adam usually wins early.**

---

### Golden rule (memorize this)

> Bad learning rate breaks good models.

> Good learning rate saves bad models.